ARCHITECTURE.md

```diff
@@ -0,0 +1,95 @@
+# ink!ternals
+
+
+This document describes the architecture of ink!. The information
+here targets those who want to understand or modify the inner
+workings of this project.
+
+In general we treat documentation as a first-class citizen.
+All crates mentioned below should be documented really well.
+You can find the crate documentation on docs.rs or for our
+`master` branch under GitHub pages. So for `ink_lang` e.g.:
+
+* [https://docs.rs/ink_lang/latest/ink_lang](https://docs.rs/ink_lang/latest/ink_lang) (latest)
+* [https://paritytech.github.io/ink/ink_lang](https://paritytech.github.io/ink/ink_lang) (`master`)
+
+ink! is composed of a number of crates that are all found in the
+`crates/` folder. On a high-level those can be grouped as:
+
+* `lang`: The ink! language itself.
+* `allocator`: The allocator for dynamic heap space allocation in a contract.
+* `engine`: An off-chain testing engine, it simulates a blockchain
+   environment and allows mocking specified conditions.
+* `env`: Exposes environmental functions, like information about the caller
+   of a contract call, getting random entropy, or e.g. self-terminating the
+	 contract.
+* `metadata`: Describes how to call a contract (it's interface and the
+   types), its storage layout, etc..
+* `prelude`: Provides an interface to typical standard library types and
+  functionality (like `vec` or `string`). Since contracts are run in a
+	`no_std` environment we provide this crate as an entrypoint for accessing
+	functionality of the standard library.
+* `primitives`: Utilities that are used internally by multiple ink! crates.
+* `storage`: The collections that are available for contract developers
+  to put in a smart contracts storage.
+
+An important thing to note is that the crates are primarily run in
+a `no_std` environment, specifically in a Wasm sandbox execution
+environment on the blockchain itself.
+Exceptions are `metadata` and `engine`, which cover use-cases that
+are only relevant off-chain.
+
+## Overview
+
+<img src="./.images/inkternals.png" alt="ink!ternals" width="800" />
```

Comment:
Don't think we need to have the IR mentioned in the diagram, that's more of an implementation detail imo

Reply:
I disagree here, I think it helps understanding what is actually happening when you build a contract.

Reply:
Well it depends on what the goal here is. If we're trying to give a high level overview the fact that we have an IR is an implementation detail. You could do this without an IR and just have "raw" proc macros